yew/src/virtual_dom/vlist.rs

```diff
@@ -78,137 +173,320 @@ impl VDiff for VList {
+        let all_keyed = self.children.len() == n_keyed_lefts;
+        if some_keyed && !all_keyed {
+            log::error!(
+                "Not all elements have keys in VList ({} keyed out of {}), this is currently not \
+                supported. Ignoring keys.",
+                n_keyed_lefts,
+                self.children.len(),
+            );
         }
 
-        // Process children
-        let lefts = self.children.iter_mut();
-        let key_count = rights.iter().filter(|r| r.key().is_some()).count();
-        let mut rights_nokeys = Vec::with_capacity(rights.len() - key_count);
-        let mut rights_lookup = HashMap::with_capacity(key_count);
-        for mut r in rights.into_iter() {
-            if let Some(key) = r.key() {
-                if rights_lookup.contains_key(key) {
-                    log::error!("Duplicate key of {}", &key);
-                    r.detach(parent);
+        // Take rights from ancestor.
+        let mut rights = match ancestor {
+            // If the ancestor is also a VList, then the right list is the
+            // previously rendered items.
+            Some(VNode::VList(vlist)) => vlist.children,
+            // Otherwise, there was a node before but it wasn't a list. Then,
+            // use the current node as a single fragment list and let the
+            // `apply` of `VNode` handle it.
+            Some(vnode) if !some_keyed => vec![vnode],
+            // Otherwise, we don't reuse it, as the chance that the element
+            // is keyed and present in left is almost null.
+            Some(mut vnode) => {
+                vnode.detach(parent);
+                vec![]
+            }
+            None => vec![],
+        };
+
+        // Collect the right/ancestor keyed elements in a hashmap.
+        let mut rights_lookup = rights
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, vnode)| vnode.key().as_ref().map(|key| (key.clone(), idx)))
+            .collect::<HashMap<_, _>>();
+
+        // Determine which algorithm we use. If there are some keys, but not all
+        // the elements are keyed, we consider it a degenerated case and use the
+        // non-keyed algorithm. Importantly, don't use the keyed algorithm if
+        // rights is empty (no need).
+        let all_rights_have_keys = rights_lookup.len() == rights.len();
+        let use_keyed_algorithm =
+            some_keyed && all_keyed && all_rights_have_keys && !rights.is_empty();
+
+        // The algorithms are different when there are keys, because it is more
+        // expensive and less frequent.
+        if !use_keyed_algorithm {
+            let mut rights = rights.into_iter();
+            for left in self.children.iter_mut() {
+                // Discard the returned node, as we are only interested in the
+                // next sibling, not the previous one.
+                let _ = left.apply(parent_scope, parent, None, rights.next());
```

Comment:
Since `rights.next()` could be `None`, we need to pass `next_sibling` instead of `None` here.

Reply:
I don't think so, and this makes me think that we should change the signature of `apply`. Here there are basically two situations: a) `rights.next()` is `Some(_)`, so we mutate it to become `left` and don't need to use `next_sibling` or b) `rights.next()` is `None` then indeed we need `next_sibling`, but if `rights` is exhausted that means that we are at the end of the list, so there are no next siblings. This is why I am passing `None` and discarding the returned `Node`.

Reply:
Right but this `VList` might have a next sibling so we can't just append to `parent`. Right now you are ignoring `next_sibling` by naming it `_next_sibling`, we should be using it

Reply:
I hadn't thought of that at all.

Reply:
This should now be fixed, I wrote two tests to check it.